File: components/elements/select_element.cpp
#include "select_element.h"

#include <limits>

namespace esphome::elements {

static const char *const SELECT_ELEMENT_TAG = "elements.select";

// ---------------------------------------------------------------------------
// SelectElement
// ---------------------------------------------------------------------------

void SelectElement::visit_children(const std::function<void(Element *, bool)> &fn) {
  for (int i = 0; i < (int) elements_.size(); i++) {
    fn(elements_[i], i == index_);
  }
}

void SelectElement::go_to(int index) {
  if (index != index_) {
    int from = index_;
    index_ = index;
    on_change(from, index);
  }
}

void SelectElement::on_change(int from_index, int to_index) {
  ESP_LOGI(SELECT_ELEMENT_TAG, "Triggering `on_change` for %s (%p): %i → %i", get_type_name(), this, from_index,
           to_index);
  on_change_callbacks_.call(this, from_index, to_index);
}

// ---------------------------------------------------------------------------
// PriorityElement
// ---------------------------------------------------------------------------

static const char *const PRIORITY_ELEMENT_TAG = "elements.priority";

int PriorityElement::find_active_index_() const {
  for (int i = 0; i < (int) elements_.size(); i++) {
    if (elements_[i]->is_active()) {
      return i;
    }
  }
  return -1;
}

void PriorityElement::update_state() {
  ContainerElement::update_state();
  go_to(find_active_index_());
}

void PriorityElement::draw(display::Display &display) {
  if (index_ != -1) {
    elements_[index_]->draw(display);
  }
}

// ---------------------------------------------------------------------------
// RandomElement
// ---------------------------------------------------------------------------

static const char *const RANDOM_ELEMENT_TAG = "elements.random";

void RandomElement::update_state() {
  ContainerElement::update_state();
  if (index_ == -1) {
    next();
  }
}

void RandomElement::draw(display::Display &display) {
  if (index_ != -1) {
    elements_[index_]->draw(display);
  }
}

void RandomElement::update_history_() {
  if (index_ != -1) {
    history_.push_back(index_);
    while (history_.size() > 2 * elements_.size()) {
      history_.pop_front();
    }
  }
}

void RandomElement::prev() {
  while (!history_.empty()) {
    int index = history_.back();
    history_.pop_back();
    if (elements_[index]->is_active()) {
      go_to(index);
      return;
    }
  }
}

void RandomElement::next() {
  std::vector<int> candidates;
  int min = std::numeric_limits<int>::max();
  for (int index = 0; index < (int) elements_.size(); index++) {
    if (!elements_[index]->is_active())
      continue;
    int count = 0;
    for (int hist_index : history_) {
      if (hist_index == index) {
        count++;
      }
    }
    if (count < min) {
      min = count;
      candidates.clear();
      candidates.push_back(index);
    } else if (count == min) {
      candidates.push_back(index);
    }
  }
  if (!candidates.empty()) {
    int index = candidates[get_component()->get_current_ms() % candidates.size()];
    update_history_();
    go_to(index);
  }
}

// ---------------------------------------------------------------------------
// SequenceElement
// ---------------------------------------------------------------------------

static const char *const SEQUENCE_ELEMENT_TAG = "elements.sequence";

void SequenceElement::update_state() {
  ContainerElement::update_state();
  if (index_ == -1) {
    next();
  }
}

void SequenceElement::draw(display::Display &display) {
  if (index_ != -1) {
    elements_[index_]->draw(display);
  }
}

void SequenceElement::prev() {
  int count = elements_.size();
  for (int offset = 1; offset < count; offset++) {
    int new_index = (index_ - offset + count) % count;
    if (elements_[new_index]->is_active()) {
      go_to(new_index);
      return;
    }
  }
}

void SequenceElement::next() {
  int count = elements_.size();
  for (int offset = 1; offset < count; offset++) {
    int new_index = (index_ + offset) % count;
    if (elements_[new_index]->is_active()) {
      go_to(new_index);
      return;
    }
  }
}

}  // namespace esphome::elements

File: components/elements/select_element.py
import esphome.config_validation as cv
import esphome.codegen as cg
import esphome.automation as automation
from esphome.const import CONF_ID, CONF_TRIGGER_ID

from . import shared
from . import element
from . import element_registry

CONF_ON_CHANGE = "on_change"
CONF_ELEMENTS = "elements"

# Abstract Select Element

SelectElement = shared.elements_ns.class_("SelectElement", element.Element)
SelectElementRef = SelectElement.operator("ref")
SelectElementChangeTrigger = shared.elements_ns.class_(
    "SelectElementChangeTrigger", automation.Trigger.template(cg.int_, cg.int_)
)

SELECT_ELEMENT_SCHEMA = element.ELEMENT_SCHEMA.extend(
    {
        cv.Required(CONF_ELEMENTS): cv.All(
            cv.ensure_list(element_registry.TYPED_ELEMENT_SCHEMA),
            cv.Length(min=1),
        ),
        cv.Optional(CONF_ON_CHANGE): automation.validate_automation(
            {
                cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(
                    SelectElementChangeTrigger
                ),
            }
        ),
    }
)


async def select_element_to_code(config, component, parent):
    var = await element.element_to_code(config, component, parent)
    if elements_configs := config.get(CONF_ELEMENTS):
        for conf in elements_configs:
            element_var = await element_registry.typed_element_to_code(
                conf, component, var
            )
            cg.add(var.add_element(element_var))

    for conf in config.get(CONF_ON_CHANGE, []):
        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
        await automation.build_automation(
            trigger, [(cg.int_, "from_index"), (cg.int_, "to_index")], conf
        )
    return var


@automation.register_action(
    "elements.next",
    shared.elements_ns.class_("SelectNextAction", automation.Action),
    automation.maybe_simple_id({cv.Required(CONF_ID): cv.use_id(SelectElement)}),
    synchronous=True,
)
async def elements_next_to_code(config, action_id, template_arg, args):
    paren = await cg.get_variable(config[CONF_ID])
    return cg.new_Pvariable(action_id, template_arg, paren)


@automation.register_action(
    "elements.prev",
    shared.elements_ns.class_("SelectPrevAction", automation.Action),
    automation.maybe_simple_id({cv.Required(CONF_ID): cv.use_id(SelectElement)}),
    synchronous=True,
)
async def elements_prev_to_code(config, action_id, template_arg, args):
    paren = await cg.get_variable(config[CONF_ID])
    return cg.new_Pvariable(action_id, template_arg, paren)


# Priority Element

PriorityElement = shared.elements_ns.class_("PriorityElement", SelectElement)

PRIORITY_ELEMENT_SCHEMA = SELECT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(PriorityElement),
    }
)

element_registry.register_element(
    "priority", PRIORITY_ELEMENT_SCHEMA, select_element_to_code
)

# Random Element

RandomElement = shared.elements_ns.class_("RandomElement", SelectElement)

RANDOM_ELEMENT_SCHEMA = SELECT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(RandomElement),
    }
)

element_registry.register_element(
    "random", RANDOM_ELEMENT_SCHEMA, select_element_to_code
)

# Sequence Element

SequenceElement = shared.elements_ns.class_("SequenceElement", SelectElement)

SEQUENCE_ELEMENT_SCHEMA = SELECT_ELEMENT_SCHEMA.extend(
    {
        cv.GenerateID(CONF_ID): cv.declare_id(SequenceElement),
    }
)

element_registry.register_element(
    "sequence", SEQUENCE_ELEMENT_SCHEMA, select_element_to_code
)
